### ISSUE TITLE
I have a similar project. Considering combing resources?

### ISSUE THREAD

**Issue by tamb** [torch2424/pubx#1]
I have a similar project called `substate`.

I'm considering rebranding it and deprecating he current package. 

Would you be interested in collaborating?

https://github.com/tamb/substate

**Comment by torch2424** [comment]
Hello @tamb :smile: Cool avatar! I love regular show haha!

Yes! So I am totally open to collaborating. That being said, UIve honestly put this project of mine on the back burner for now, so I haven't been active on it. I've been really focused on my WebAssembly projects. So if the collaboarting is more chatting or reviewing PRs I'm totally down! :smile: 

Let me know what you think! :+1:

**Comment by tamb** [comment]
I'm totally down for that.  I gotta check your source code.  I'm going for a a pub/sub based state management with optional deep cloning.  I found that normalizing state data is sometimes not the way to go.  So I wanted the options to do that.  How about by Friday we have a couple notes on each others projects??

That sound good?

"Zingooooooooo!"

**Comment by tamb** [comment]
So my initial critique is that this isn't a true pubsub.  You there's no need to generate ids for the callbacks.  You can and probably should store them in a separate object where each key is the `action` or `event` name you subscribe to and then there's an array/Set of functions to call.

Your pattern relies on state values existing state fields for any events to be fired.  It also is not functional and mutates the state object directly.  

You can use `Object.assign` to clone an object, but again, you're restricting state to being flat.  

These are the major differences between your lib and mine.

Would love some feedback!  

OTherwise it's super concise and easy to see how state is managed.

**Comment by torch2424** [comment]
@tamb 

> I'm totally down for that. I gotta check your source code. I'm going for a a pub/sub based state management with optional deep cloning. I found that normalizing state data is sometimes not the way to go. So I wanted the options to do that. How about by Friday we have a couple notes on each others projects??

Oops! Sorry I just saw this! been super busy juggling so many open source things as of lately :disappointed: 

Deep cloning sounds good! If I remember correctly, I wasn't doing any serious deep cloning. Just some top level stuff :smile: 

> So my initial critique is that this isn't a true pubsub. You there's no need to generate ids for the callbacks. You can and probably should store them in a separate object where each key is the action or event name you subscribe to and then there's an array/Set of functions to call.

Oh yeah, definitely don't need ids. I wrote this like two years ago (was originally in one of my repos, but I pulled it out), and honestly it worked for what I needed it for, and haven't really updated or anything beyond what I needed for my projects. So totally would not be surprised if it is not true pub sub :joy: I can totally update the description if it falls under something else :smile: 

> Your pattern relies on state values existing state fields for any events to be fired. It also is not functional and mutates the state object directly.

Yeah totally, so I do want a "central state". Perhpas the pub/sub is a bit misleading for my library. I built this just as an easy way for me to maintain state in a single place for my [WasmBoy](https://github.com/torch2424/wasmboy) and [VaporBoy](https://github.com/torch2424/vaporBoy) projects. 

Taking a step back, what I really wanted was a: Central state that I can subscribe to certain keys on that state, and react whenever a new value to said key is published.

So perhpas it s a pub/sub-like state management? :thinking: 

> These are the major differences between your lib and mine.

Yeah totally, looking over yours I am definitely noticing that. Perhaps it comes down to keeping the projects seperate, and explaining what mine is trying to do a bit better? :thinking: 

> Would love some feedback!

I do appreciate how much more functional, and configurable yours is! :smile: Great work! :+1: 

> OTherwise it's super concise and easy to see how state is managed.

Thank you! I appreciate the feedback!
